### ISSUE TITLE
Stacking Contract Signature and Input Mismatch

### ISSUE THREAD

**Issue by rachit2501** [blockstack/docs#1029]
- The stacking contract signature and inpute values are mismatched in order. Hard to understand which one is related to which one. 
- No example on how to modify pox-addr to supoorted format.

**Comment by agraebe** [comment]
Can you please add the concrete mismatches? In my understanding it is hard to read but accurate.

The references are generated from the contract itself and I believe every change would trigger an automatic update in the docs

**Comment by rachit2501** [comment]
> Can you please add the concrete mismatches? In my understanding it is hard to read but accurate.
> 
> The references are generated from the contract itself and I believe every change would trigger an automatic update in the docs.
> 
> As for the pox address, we absolutely have to add information about that. We also should complete the delegation steps in the tutorial with sample code to make it easier. Any help is very welcome

Sure. For eg.:
![delegatestx](https://docs.blockstack.org/en-US/references/stacking-contract#delegate-stx)
My options are:

```
const options = {
    contractAddress: "ST000000000000000000002AMW42H",
    contractName: "pox",
    functionName: "delegate-stx",
    functionArgs: [
      uintCV(amountSTX),
      standardPrincipalCV("ST3K2B2FH1AYXD26WV6YZY4DAA82AZNK967BNB9BK"),
      poxAddressCV,
    ],

    appDetails: {
      name: "StakedStats",
      icon: { logo },
    },
    finished: (data) => {
      console.log("TX ID:", data.txId);
      console.log("Raw TX:", data.txRaw);
    },
  };
```
I am getting bad argument error.

Also:
![](https://docs.blockstack.org/en-US/references/stacking-contract#delegate-stack-stx). There are two uint in the end, but only one uint value after pox.

**Comment by agraebe** [comment]
TL;DR: The ui of the docs seems confusing but the content is correct here. You are tripped up by the fact that there's no tutorial or guide - no explanation _how_ to use what is documented in the refs, I'd suggest we close this out and open an issue for tutorial and guide for stacking delegation.

----

The delegate method is documented as:

```
Signature: (delegate-stx amount-ustx delegate-to until-burn-ht pox-addr)
Input: uint, principal, (optional uint), (optional (tuple (hashbytes (buff 20)) (version (buff 1))))
```

You use uint, principal, and what I assume is a tuple. you need to add `nonceCV` for the optional param (the 3rd in the list for `until-burn-ht`).

----

The link on the bottom is for `delegate-stack-stx`, which is defined as:

```
Signature: (delegate-stack-stx stacker amount-ustx pox-addr start-burn-ht lock-period)
Input: principal, uint, (tuple (hashbytes (buff 20)) (version (buff 1))), uint, unit
```

There are two uints at the end of the signature: `start-burn-ht` and `lock-period`.

**Comment by rachit2501** [comment]
> TL;DR: The ui of the docs seems confusing but the content is correct here. You are tripped up by the fact that there's no tutorial or guide - no explanation _how_ to use what is documented in the refs, I'd suggest we close this out and open an issue for tutorial and guide for stacking delegation.
> 
> The delegate method is documented as:
> 
> ```
> Signature: (delegate-stx amount-ustx delegate-to until-burn-ht pox-addr)
> Input: uint, principal, (optional uint), (optional (tuple (hashbytes (buff 20)) (version (buff 1))))
> ```
> 
> You use uint, principal, and what I assume is a tuple. you need to add `noneCV` for the optional param (the 3rd in the list for `until-burn-ht`).
> 
> The link on the bottom is for `delegate-stack-stx`, which is defined as:
> 
> ```
> Signature: (delegate-stack-stx stacker amount-ustx pox-addr start-burn-ht lock-period)
> Input: principal, uint, (tuple (hashbytes (buff 20)) (version (buff 1))), uint, unit
> ```
> 
> There are two uints at the end of the signature: `start-burn-ht` and `lock-period`.

Ok. So I figured out the issue,
 (optional (tuple (hashbytes (buff 20)) (version (buff 1))))
A guide on how to create this datatype, needs to be documented.

```
const { hashMode, data } = decodeBtcAddress(poxAddr);
  const hashModeBuffer = bufferCV(new BN(hashMode, 10).toArrayLike(Buffer));
  const hashbytes = bufferCV(data);
  const poxAddressCV = tupleCV({
    hashbytes,
    version: hashModeBuffer,
  });
```
This is the code i made by reading the `@stacks/transaction` library code. And seems like this is the argument giving me `"BAD ARGUMENT" `error.

**Comment by agraebe** [comment]
Yeah, agree with your comment that a) the references are not so readable and b) there is missing sample code for this call.

I have some working code in a sample app, hoping it helps you (until we get better tooling, docs, and samples for this):

---

Install deps:

```shell
npm install @stacks/transactions @stacks/network bitcoinjs-lib bn.js
```

Make sure to get the latest versions of the stacks libs because the mainnet node URLs were updated recently.

Sample code:

```js
const {
  AddressHashMode,
  bufferCV,
  tupleCV,
  standardPrincipalCV,
  uintCV,
  makeContractCall,
  broadcastTransaction,
} = require("@stacks/transactions");
const { StacksMainnet } = require("@stacks/network");
const { address } = require("bitcoinjs-lib");
const BN = require("bn.js");

const network = new StacksMainnet();

const BLOCK_HEIGHT = 1903893;
const MINIMUM_USTX = 85432771473374;

async function stackTokensForUser(address, senderKey, btcAddress) {
  const { hashMode, data } = decodeBtcAddress(btcAddress);
  const hashModeBuffer = bufferCV(new BN(hashMode, 10).toBuffer());
  const hashbytes = bufferCV(data);
  const poxAddressCV = tupleCV({
    hashbytes,
    version: hashModeBuffer,
  });

  const txOptions = {
    contractAddress: "ST000000000000000000002AMW42H",
    contractName: "pox",
    functionName: "delegate-stack-stx",
    functionArgs: [
      standardPrincipalCV(address),
      uintCV(MINIMUM_USTX * 1.1), // amount+10%
      poxAddressCV,
      uintCV(BLOCK_HEIGHT + 3), //start-burn-ht+3
      uintCV(1), //lock-period
    ],
    senderKey,
    validateWithAbi: true,
    network,
  };

  const transaction = await makeContractCall(txOptions);

  return broadcastTransaction(transaction, network);
}

function decodeBtcAddress(btcAddress) {
  const hashMode = getAddressHashMode(btcAddress);
  if (btcAddress.startsWith("bc1") || btcAddress.startsWith("tb1")) {
    const { data } = address.fromBech32(btcAddress);
    return {
      hashMode,
      data,
    };
  } else {
    const { hash } = address.fromBase58Check(btcAddress);
    return {
      hashMode,
      data: hash,
    };
  }
}

function getAddressHashMode(btcAddress) {
  if (btcAddress.startsWith("bc1") || btcAddress.startsWith("tb1")) {
    const { data } = address.fromBech32(btcAddress);
    if (data.length === 32) {
      return AddressHashMode.SerializeP2WSH;
    } else {
      return AddressHashMode.SerializeP2WPKH;
    }
  } else {
    const { version } = address.fromBase58Check(btcAddress);
    switch (version) {
      case 0:
        return AddressHashMode.SerializeP2PKH;
      case 111:
        return AddressHashMode.SerializeP2PKH;
      case 5:
        return AddressHashMode.SerializeP2SH;
      case 196:
        return AddressHashMode.SerializeP2SH;
      default:
        throw new Error("Invalid pox address version");
    }
  }
}
```
